src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs

```diff
@@ -204,69 +217,114 @@ private void Shutdown(Exception ex = null)
+                        call.Complete(ex);
                     }
+
+                    // This will cancel the completion if nobody has done so already.
+                    call.Dispose();
                 }
                 _pendingCalls.Clear();
             }
         }
 
-        private void DispatchInvocation(InvocationDescriptor invocationDescriptor, CancellationToken cancellationToken)
+        private void DispatchInvocation(InvocationMessage invocation, CancellationToken cancellationToken)
         {
             // Find the handler
-            if (!_handlers.TryGetValue(invocationDescriptor.Method, out InvocationHandler handler))
+            if (!_handlers.TryGetValue(invocation.Target, out InvocationHandler handler))
             {
-                _logger.LogWarning("Failed to find handler for '{0}' method", invocationDescriptor.Method);
+                _logger.LogWarning("Failed to find handler for '{0}' method", invocation.Target);
                 return;
             }
 
             // TODO: Return values
             // TODO: Dispatch to a sync context to ensure we aren't blocking this loop.
-            handler.Handler(invocationDescriptor.Arguments);
+            handler.Handler(invocation.Arguments);
         }
 
-        private void DispatchInvocationResult(InvocationResultDescriptor result, InvocationRequest irq, CancellationToken cancellationToken)
+        private void DispatchInvocationResult(StreamItemMessage result, InvocationRequest irq, CancellationToken cancellationToken)
         {
-            _logger.LogInformation("Received Result for Invocation #{0}", result.Id);
+            _logger.LogInformation("Received Result for Invocation #{0}", result.InvocationId);
 
             if (cancellationToken.IsCancellationRequested)
             {
                 return;
             }
 
-            Debug.Assert(irq.Completion != null, "Didn't properly capture InvocationRequest in callback for ReadInvocationResultDescriptorAsync");
-
             // If the invocation hasn't been cancelled, dispatch the result
             if (!irq.CancellationToken.IsCancellationRequested)
             {
-                irq.Registration.Dispose();
+                irq.ReceiveResult(result.Item);
+            }
+        }
+
+        private void DispatchInvocationCompletion(CompletionMessage completion, InvocationRequest irq, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Received Completion for Invocation #{0}", completion.InvocationId);
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return;
+            }
+
+            // If the invocation hasn't been cancelled, dispatch the completion
+            if (!irq.CancellationToken.IsCancellationRequested)
+            {
+                if (!string.IsNullOrEmpty(completion.Error))
+                {
+                    irq.Complete(new Exception(completion.Error));
```

Comment:
Given that this is generally an application-defined error (aside from a few cases), we should use `Exception` unless we want to encode an exception type in the Error frame. If we want to throw specific exceptions for our well-known errors, maybe we need to associate a numeric code or symbolic name to some errors?